File: modeling/convert.py
"""
Blender Python 스크립트 — LINK STEP → GLB 일괄 변환.

전제 조건:
  - Blender 4.2+ 설치
  - STEP 임포트 애드온 활성화 (예: "STEPper", "Free CAD STEP import")

사용:
  blender --background --python modeling/convert.py -- modeling/IRB2600

위 명령은 modeling/IRB2600/ 폴더의 *_BASE_CAD_*.STEP, *_LINK[1-6]_CAD_*.STEP 을
같은 폴더에 base.glb, link1.glb ~ link6.glb 으로 저장합니다.
"""
import bpy
import os
import sys
import re

# 인자 파싱: blender --python script.py -- <input_dir>
argv = sys.argv
if "--" in argv:
    argv = argv[argv.index("--") + 1:]
else:
    argv = []
target_dir = argv[0] if argv else "."
target_dir = os.path.abspath(target_dir)

NAME_MAP = {
    r"_BASE_CAD_": "base.glb",
    r"_LINK1_CAD_": "link1.glb",
    r"_LINK2_CAD_": "link2.glb",
    r"_LINK3_CAD_": "link3.glb",
    r"_LINK4_CAD_": "link4.glb",
    r"_LINK5_CAD_": "link5.glb",
    r"_LINK6_CAD_": "link6.glb",
}

def find_step_files(folder):
    out = []
    for f in os.listdir(folder):
        if not f.upper().endswith(".STEP"):
            continue
        for pat, glb in NAME_MAP.items():
            if re.search(pat, f, re.IGNORECASE):
                out.append((os.path.join(folder, f), os.path.join(folder, glb)))
                break
    return out

def import_step(path):
    # 다양한 STEP 임포트 애드온 시도
    if hasattr(bpy.ops, "import_step"):
        bpy.ops.import_step.step(filepath=path)
        return True
    if hasattr(bpy.ops.import_mesh, "step"):
        bpy.ops.import_mesh.step(filepath=path)
        return True
    if hasattr(bpy.ops.wm, "step_import"):
        bpy.ops.wm.step_import(filepath=path)
        return True
    raise RuntimeError("STEP 임포트 애드온이 없습니다. STEPper 또는 Free CAD STEP import를 활성화하세요.")

def main():
    files = find_step_files(target_dir)
    if not files:
        print(f"[!] {target_dir} 에서 LINK/BASE STEP 파일을 찾지 못함")
        return
    print(f"[+] {len(files)} 개 변환 예정")
    for src, dst in files:
        print(f"  {os.path.basename(src)} -> {os.path.basename(dst)}")
        bpy.ops.wm.read_factory_settings(use_empty=True)
        try:
            import_step(src)
        except Exception as e:
            print(f"    임포트 실패: {e}")
            continue
        # 단위 m, 모든 객체 선택
        bpy.ops.object.select_all(action="SELECT")
        # GLB export (Y-up = Three.js 기본)
        bpy.ops.export_scene.gltf(
            filepath=dst,
            export_format="GLB",
            export_apply=True,
            export_yup=True,
        )
    print("[+] 완료")

if __name__ == "__main__":
    main()

File: modeling/convert.mjs
// STEP → GLB 자동 변환 (Node.js, occt-import-js + gltf-transform)
// 사용: node modeling/convert.mjs modeling/IRB2600
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import occtImportJs from 'occt-import-js';
import { Document, NodeIO } from '@gltf-transform/core';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

// LINK 명명 규칙은 ABB 모델별로 다름 — `_LINK1_CAD_` 또는 `_LINK01_CAD_` 모두 매칭
const NAME_MAP = [
  [/_BASE_CAD/i,        'base.glb'],
  [/_LINK0?1_CAD/i,     'link1.glb'],
  [/_LINK0?2_CAD/i,     'link2.glb'],
  [/_LINK0?3_CAD/i,     'link3.glb'],
  [/_LINK0?4_CAD/i,     'link4.glb'],
  [/_LINK0?5_CAD/i,     'link5.glb'],
  [/_LINK0?6_CAD/i,     'link6.glb'],
];

function findFiles(dir) {
  const entries = fs.readdirSync(dir);
  const out = [];
  for (const f of entries) {
    if (!/\.step$/i.test(f)) continue;
    for (const [pat, glb] of NAME_MAP) {
      if (pat.test(f)) {
        out.push({ src: path.join(dir, f), dst: path.join(dir, glb) });
        break;
      }
    }
  }
  return out;
}

async function convertOne(occt, srcPath, dstPath) {
  const buf = fs.readFileSync(srcPath);
  const result = occt.ReadStepFile(new Uint8Array(buf), null);
  if (!result.success || !result.meshes || result.meshes.length === 0) {
    throw new Error('STEP 파싱 실패 또는 메쉬 없음');
  }
  const doc = new Document();
  const buffer = doc.createBuffer();
  const scene = doc.createScene();
  for (let i = 0; i < result.meshes.length; i++) {
    const m = result.meshes[i];
    // m.attributes.position.array = Float32Array of vertex coords (mm by default)
    // m.index.array = Uint32Array of triangle indices
    const srcPos = m.attributes.position.array;
    const positions = new Float32Array(srcPos.length);
    // mm → m 변환 + CAD Z-up → Three.js Y-up 변환 (Rx(-π/2): (x,y,z) → (x, z, -y))
    for (let j = 0; j < srcPos.length; j += 3) {
      positions[j]     =  srcPos[j]   / 1000;          // x → x
      positions[j + 1] =  srcPos[j+2] / 1000;          // z → y (height)
      positions[j + 2] = -srcPos[j+1] / 1000;          // -y → z
    }
    const indices = new Uint32Array(m.index.array);
    let normals = null;
    if (m.attributes.normal) {
      const srcN = m.attributes.normal.array;
      normals = new Float32Array(srcN.length);
      for (let j = 0; j < srcN.length; j += 3) {
        normals[j]     =  srcN[j];
        normals[j + 1] =  srcN[j+2];
        normals[j + 2] = -srcN[j+1];
      }
    }

    const posAcc = doc.createAccessor()
      .setType('VEC3')
      .setArray(positions)
      .setBuffer(buffer);
    const idxAcc = doc.createAccessor()
      .setType('SCALAR')
      .setArray(indices)
      .setBuffer(buffer);
    const prim = doc.createPrimitive()
      .setAttribute('POSITION', posAcc)
      .setIndices(idxAcc);
    if (normals) {
      const nAcc = doc.createAccessor().setType('VEC3').setArray(normals).setBuffer(buffer);
      prim.setAttribute('NORMAL', nAcc);
    }
    // 머터리얼 — ABB Orange (Pantone 158C, 약 #FF6B1F)
    const mat = doc.createMaterial(`mat_${i}`)
      .setBaseColorFactor([1.0, 0.42, 0.12, 1.0])
      .setMetallicFactor(0.2)
      .setRoughnessFactor(0.55);
    prim.setMaterial(mat);
    const mesh = doc.createMesh().addPrimitive(prim);
    const node = doc.createNode(`mesh_${i}`).setMesh(mesh);
    scene.addChild(node);
  }
  const io = new NodeIO();
  await io.write(dstPath, doc);
}

async function main() {
  const dir = process.argv[2] ? path.resolve(process.argv[2]) : process.cwd();
  if (!fs.existsSync(dir)) {
    console.error(`[!] 폴더 없음: ${dir}`);
    process.exit(1);
  }
  const files = findFiles(dir);
  if (!files.length) {
    console.error(`[!] LINK/BASE STEP 파일을 찾지 못함: ${dir}`);
    process.exit(1);
  }
  console.log(`[+] ${dir} 에서 ${files.length}개 변환 시작`);
  const occt = await occtImportJs();
  for (const { src, dst } of files) {
    try {
      console.log(`  ${path.basename(src)} → ${path.basename(dst)}`);
      await convertOne(occt, src, dst);
    } catch (e) {
      console.error(`    실패: ${e.message}`);
    }
  }
  console.log('[+] 완료');
}

main().catch(e => { console.error(e); process.exit(1); });
